bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/BKTClientImpl.kt

```diff
@@ -0,0 +1,169 @@
+  }
+
+  override fun track(goalId: String, value: Double) {
+    val user = userHolder.get()
+    val featureTag = config.featureTag
+    executor.execute {
+      component.eventInteractor.trackGoalEvent(
+        featureTag = featureTag,
+        user = user,
+        goalId = goalId,
+        value = value,
+      )
+    }
+  }
+
+  override fun currentUser(): BKTUser {
+    return userHolder.get().toBKTUser()
+  }
+
+  override fun setUserAttributes(attributes: Map<String, String>) {
+    userHolder.update { it.copy(data = attributes) }
+  }
+
+  private fun fetchEvaluationsSync(timeoutMillis: Long?): BKTException? {
+    val result = component.evaluationInteractor.fetch(user = userHolder.get(), timeoutMillis)
+
+    executor.execute {
+      val interactor = component.eventInteractor
+      when (result) {
+        is GetEvaluationsResult.Success -> {
+          interactor.trackFetchEvaluationsSuccess(
+            featureTag = result.featureTag,
+            mills = result.millis,
+            sizeByte = result.sizeByte,
+          )
+        }
+        is GetEvaluationsResult.Failure -> {
+          interactor.trackFetchEvaluationsFailure(
+            featureTag = result.featureTag,
+            error = result.error,
+          )
+        }
+      }
+    }
+
+    return when (result) {
+      is GetEvaluationsResult.Success -> null
+      is GetEvaluationsResult.Failure -> result.error
+    }
+  }
+
+  override fun fetchEvaluations(timeoutMillis: Long?): Future<BKTException?> {
+    return executor.submit<BKTException?> {
+      fetchEvaluationsSync(timeoutMillis)
+    }
+  }
+
+  // should we return Future?
+  override fun flush() {
+    executor.execute {
+      component.eventInteractor.sendEvents(force = true)
+    }
+  }
+
+  override fun evaluationDetails(featureId: String): BKTEvaluation? {
+    val raw = component.evaluationInteractor.getLatest(userHolder.userId, featureId) ?: return null
+
+    return BKTEvaluation(
+      id = raw.id,
+      featureId = raw.feature_id,
+      featureVersion = raw.feature_version,
+      userId = raw.user_id,
+      variationId = raw.variation_id,
+      variationValue = raw.variation_value,
+      reason = raw.reason.type.value,
+    )
+  }
+
+  private fun refreshCache() {
+    component.evaluationInteractor.refreshCache(userHolder.userId)
+    component.eventInteractor.refreshCache()
+  }
+
+  internal fun initializeInternal(timeoutMillis: Long): Future<BKTException?> {
+    return executor.submit<BKTException?> {
+      refreshCache()
+      fetchEvaluationsSync(timeoutMillis)
+    }
+  }
+
+  private inline fun <reified T : Any> getVariationValue(featureId: String, defaultValue: T): T {
+    logd { "BKTClient.getVariation(featureId = $featureId, defaultValue = $defaultValue) called" }
+
+    val raw = component.evaluationInteractor.getLatest(userHolder.userId, featureId)
+
+    val user = userHolder.get()
+    val featureTag = config.featureTag
+    if (raw != null) {
```

Comment:
If `raw` can be null, we should also check here.

https://github.com/ca-dp/bucketeer-android-sdk/pull/68/files#diff-933843224cca039e7d3bbd7550c39d420051db3ac75ba15284cb3adca143ff45R167

Reply:
```kotlin
fun <reified T : Any> Evaluation?.getVariationValue(defaultValue: T): T
```
https://github.com/ca-dp/bucketeer-android-sdk/pull/68/files#diff-87f9b54a055988f65dace3e487714328a214f8cef5a81d84076c4594b76b6953R8

receiver type of `getVariationValue()` is `Evaluation?`, so null check is handled inside the function.

Reply:
Oh ok, thanks!